source/http_connection.c

```diff
@@ -158,59 +197,76 @@ napi_value aws_napi_http_connection_new(napi_env env, napi_callback_info info) {
         goto argument_error;
     }
 
-    struct aws_tls_ctx *tls_ctx = NULL;
-    if (napi_get_value_external(env, node_args[6], (void **)&tls_ctx)) {
-        napi_throw_error(env, NULL, "Failed to extract tls_ctx from external");
-        goto argument_error;
+    if (!aws_napi_is_null_or_undefined(env, node_args[6])) {
+        if (napi_get_value_external(env, node_args[6], (void **)&tls_ctx)) {
+            napi_throw_error(env, NULL, "Failed to extract tls_ctx from external");
+            goto argument_error;
+        }
     }
 
     /* create node external to hold the connection wrapper, cleanup is required from here on out */
-    struct http_nodejs_connection *node_connection =
-        aws_mem_calloc(allocator, 1, sizeof(struct http_nodejs_connection));
-    if (!node_connection) {
+    struct http_connection_binding *binding = aws_mem_calloc(allocator, 1, sizeof(struct http_connection_binding));
+    if (!binding) {
         aws_napi_throw_last_error(env);
         goto alloc_failed;
     }
 
     napi_value node_external;
-    if (napi_create_external(env, node_connection, NULL, NULL, &node_external)) {
-        napi_throw_error(env, NULL, "Failed to create napi external for http_nodejs_connection");
+    if (napi_create_external(env, binding, NULL, NULL, &node_external)) {
+        napi_throw_error(env, NULL, "Failed to create napi external for http_connection_binding");
         goto create_external_failed;
     }
 
-    node_connection->uv_context = aws_uv_context_get_default();
-    aws_uv_context_acquire(node_connection->uv_context, env);
-    node_connection->on_setup = on_connection_setup;
-    node_connection->on_shutdown = on_connection_shutdown;
+    if (napi_create_reference(env, node_external, 1, &binding->node_external)) {
+        napi_throw_error(env, NULL, "Failed to reference node_external");
+        goto create_external_failed;
+    }
 
+    binding->env = env;
+    binding->uv_context = aws_uv_context_get_default();
+    aws_uv_context_acquire(binding->uv_context, env);
```

Comment:
if `aws_uv_context_acquire()` can't fail, let's make it a `void` function

Reply:
So, it can fail, let me see what I can do about that.